### DISCUSSION TITLE
Are there any samples for getting started with UWP and xbox UWP?

### DISCUSSION THREAD

**Discussion by BradChase2011** [amerkoleci/Vortice.Windows#414] [Q&A]
We currently have everything up and running using GLFW and Vortice.Windows and it runs extremely well.  That said, we would like to also be able to run on Windows UWP and XBOX UWP.  Are there any samples/examples anywhere on setting that up?  I have seen a little bit stated of using a swap chain control in the xaml but am pretty confused?  I also noticed there is the Vortice.WinUI library, is that used for UWP and xbox uwp?  Basically we are looking for anything that has some startup code for UWP.  Thanks much!

**Comment by amerkoleci** [comment]
Hi,
As far as I know you can easily use Vortice.Windows in WinU3 applications, specially with SwapChainPanel.

UWP and the Xbox path required the legacy netstandard2.0 version of Vortice.Windows, which actually isn't the case anymore, would be interesting to explore the https://github.com/microsoft/CsWinRT if it makes still possible. 

Cheers!

**Reply by BradChase2011** [reply]
SUPER COOL!  I had no idea they were even working on a C# RT.  I will look into it and try to learn about it.  We are in the midst of trying to get a netstandard2.0 target framework for our engine and game but man o man going backwards is painful.  It would be a dream if the C# RT works.  Have you looked into it at all?  I noticed you said "which actually isn't the case anymore", did you mean there is another path rather than netstandard2.0 or are you saying the C# RT would be that path?

We want to go both ways Creator program and gdxk.  Alot of devs simply cant get access to ID@Xbox program and it really hampers engine usage.

That all said, do you or anyone know of a simple example to use the UWP swap chain?  I personally have never used the UWP swap chain stuff and am kinda blanking out on how to hook it up.  We did WinForms and ended up switching to GLFW to keep the code closer to our vulkan renderer.

Thanks a bunch!

**Reply by BradChase2011** [reply]
I looked at the latest code and it appears the netstandard2.0 was taken out, is there a reason for that?  Like another path forward or should we reference older packages?

**Reply by amerkoleci** [reply]
You can reference old versions or try the CsWinRT with net7.0 or higher

**Reply by BradChase2011** [reply]
I would love to try CsWinRT but don't know much about it.  Is it something I can just overlay on the vortice.windows libraries or does it need built in as an endpoint/framework to the vortice libraries?  I went ahead and tried the older version of the Win32 netstandard2.0 and didn't realize how different it is than the vortice windows libraries.  Basically have to change all the code.  I think I'll take a break and do some research into the CsWinRT.

**Reply by amerkoleci** [reply]
Right, makes sense to focus on net8 and further.

Are you developing a game engine?

**Reply by BradChase2011** [reply]
Not sure that is what I meant, I would love to try the CsWinRT route to get it working on xbox, just not sure if I need to modify the vortice libraries or wrap them when I call them from the engine side.  Yea we are currently developing a game engine and are in the process of moving it over to the xbox for the community and gdxk targets.  We have moved all of our code over to .NET 8 at this point (and LOVING every minute of it) but unfortunately Microsoft has held the xbox back so much that its forcing us to make new targets using UWP and netstandard2.0.  We used to use SharpDX before the vortice libraries and thats actually alot closer to Vortice.Windows that the Vortice.Win32, unfortunately its missing alot of the newer features we need but it still targets netstandard2.0 (1.1 I believe).

**Reply by BradChase2011** [reply]
Well to update, we have the engine running under community Xbox using SharpDX and netstandard2.0.  Now to do some research and find out how to use vortice.

**Reply by amerkoleci** [reply]
I know FNA has C# port of net8 runtime with native AOT, try to reach on their discord and ask about.

**Reply by BradChase2011** [reply]
Sounds good, I skimmed their library a few days ago but now that we atleast know the core parts all run on Xbox we can do some deeper digging. Good recommendation, thanks much.

**Reply by BradChase2011** [reply]
The XBox performance under game mode through UWP creators side is pretty darn slow.  Comparable machine will get around 950fps under a test using .NET8 while the xbox gets around 115fps.  Forgot how slow the old .net was.  So we tried to go the cswinrt route but ran into a road block.  From everything I can find, you cannot have any unsealed classes.  That is an obvious deal breaker for us.  Apparently the requirement is fine if you go from C++ to C#, but not C# to C#.  Maybe there is another way?

EDIT:  I just reread your old post about NativeAOT, I looked into it a bit in their code but will have to do more digging.  I saw a post by one of its developers stating that they use their "Core" library for xbox which I believe is either .net7 or 8, but they dont support UWP any longer and only use it for the xdk if memory serves.

**Reply by BradChase2011** [reply]
Ok, well we just made a wrapper around our SharpDX port but instead using NativeAOT and we are now getting 400fps in our test from the 115fps using netstandard.  This is in the community program with a UWP entry point and a wrapper around our .NET 8 libraries using NativeAOT.

Quite the difference in speed!  So indeed you were correct @amerkoleci, NativeAOT does indeed work and its 4x faster!  Our next step is to incorp our DirectX 12 libraries using your Vortice.Windows and see what the speed is there.  Cant complain given its on an xbox series s with community which is basically half the cpu and gpu available to the id program.

**Reply by BradChase2011** [reply]
Got it moved over to using Vortice and there are some issues I had.  First off, the FPS was around 420fps, so slightly better than the SharpDX libraries.

The two main issues I had are creating the swap chain, because the .net 8 NativeAOT is hidden behind a wrapper, we cant send a SwapChainPanel over from UWP, instead we have to send a pointer.  From what I can tell, we cant compile it using the Vortice.DXGI library under UWP so I had to include the SharpDX.DXGI library to get the nativepointer...

`
            SwapChainPanel SwapElement = new SwapChainPanel() { HorizontalAlignment = HorizontalAlignment.Stretch, VerticalAlignment = VerticalAlignment.Stretch };
            Content = SwapElement;

            ISwapChainPanelNative nativeObject = ComObject.As<ISwapChainPanelNative>(SwapElement);
            Context = Startup(nativeObject.NativePointer);`

Startup being the static call into the wrapper.  Once into the .NET 8 side, we can recreate the swap chain like so:

`            IDXGISwapChain1 tempSwapChain = factory.CreateSwapChainForComposition(_GraphicsQueue, swapChainDesc1);
            nint outPtr;
            tempSwapChain.QueryInterface(typeof(IDXGISwapChain3).GetTypeInfo().GUID, out outPtr);
            _SwapChain = (IDXGISwapChain3)Activator.CreateInstance(typeof(IDXGISwapChain3), outPtr);

            nint outPtr2;
            ComObject comObject = new ComObject(NativeWindow);
            comObject.QueryInterface(typeof(ISwapChainPanelNative).GetTypeInfo().GUID, out outPtr2);
            ISwapChainPanelNative nativeObject = (ISwapChainPanelNative)Activator.CreateInstance(typeof(ISwapChainPanelNative), outPtr2);
            nativeObject.SetSwapChain(_SwapChain);
            tempSwapChain.Dispose();
`

So thats the first issue, not being able to get the handle using the DXGI library.

The second issue is with XInput.  It couldnt find the xinput libraries using the NativeLibrary.TryLoad.  We changed it temporarily on ours instead to:
`
        private static nint LoadXInputLibrary(out XInputVersion version)
        {
            nint libraryHandle = LoadLibrary("xinput1_4.dll");
            if(libraryHandle != nint.Zero )
            {
                version = XInputVersion.Version14;
                return libraryHandle;
            }
            libraryHandle = LoadLibrary("xinput1_3.dll");
            if (libraryHandle != nint.Zero)
            {
                version = XInputVersion.Version13;
                return libraryHandle;
            }
            libraryHandle = LoadLibrary("xinput9_1_0.dll");
            if (libraryHandle != nint.Zero)
            {
                version = XInputVersion.Version910;
                return libraryHandle;
            }

            //version = XInputVersion.Version14;
            version = XInputVersion.Invalid;
            return 0;
        }

        [DllImport("kernel32", CharSet = CharSet.Unicode, EntryPoint = "LoadLibrary", SetLastError = true)]
        private static extern IntPtr LoadLibrary(string lpFileName);
`


I am happy with where it is so far considering all the limitations.  If those two work arounds werent needed, it would be almost the exact same code that we use for our DirectX 12 target which we open up to the likes of GLFW.

Thanks much!

**Reply by amerkoleci** [reply]
Not sure I correctly understand what the issues are, I had no issues on using ISwapChainPanelNative which are mapped under Vortice.DXGI (https://github.com/amerkoleci/Vortice.Windows/blob/main/src/Vortice.DXGI/Mappings.xml#L19), Xinput was fixed by using LoadLibrary and not using NativeLibrary right?

**Reply by BradChase2011** [reply]
It runs great, no issue there.  The idea would be nice to send in the SwapChainPanel (UWP's) pointer into the wrapped .NET 8 library.  The problem is visual studio wont include the package Vortice.DXGI into a UWP app.  We could do this manually or simply use SharpDX.DXGI which is what we are temporarily using.  The swap panel works and we can build the entire chain, didnt know if you would want to enable your DXGI to be used in the future incase anyone else wants to build for UWP.  Its a weird scenario for sure.

The second was XInput, I could not get it to find the libraries even though they exist on both the windows box and the xbox.  I simply changed out the NativeLibrary for the kernel32 LoadLibrary and that found them.  I dont use NativeLibrary so am unsure if it just needs some sort of initialization or hinting but all the XInput stuff works great once libraries are found.

Those were the only two issues getting this all running under NativeAOT and UWP.  Everything else went very well!
